docs: add architecture doc (#1026)
matter-labs/foundry-zksync: docs-zksync/architecture/architecture.md

Upon execution, we simply convert the ZKsync VM result into what revm expects and skip the opcode execution within the revm itself. The entire execution happens in `zksync_try_create` and `zksync_try_call` of the `ZksyncCheatcodeInspectorStrategyRunner`.

More details are provided in the [book](https://foundry-book.zksync.io/zksync-specifics/execution-overview).

### Storage Translations
In order for the revm storage to be correctly utilized in ZKsync VM, we perform storage migrations to translate EVM information like balance, nonce, code storage, etc. to their respective ZKsync VM counterparts. This step is only executed once per VM switch in both `EVM <> ZKsync VM` and also in `ZKsync VM <> EVM` scenarios.

![storage](./storage.png)

⚠️ Note: Almost all cheatcodes are thus supported out of the box since we let foundry handle cheatcode application before the transaction is intercepted. Certain cheatcodes (e.g. `expectEmit`) however require additional work to be done as they must follow into the ZKsync VM.

⚠️ Note: The presence of `Cheatcode` inspector is required within a runner (specified in the builder config) to support running transactions in ZKsync VM.

## `forge script`

> Script scripts are generally executed in two stages:

Reviewer: ```suggestion
Forge scripts are generally executed in three stages:
```

> 1. Initial Execution
> 2. Broadcast

Reviewer: ```suggestion
1. Initial Execution
2. Simulation
3. Broadcast
```

> During initial execution we execute the script just like a test, while we record individual transactions as they occur using the `vm.startBroadcast() / vm.stopBroadcast()` cheatcodes. These transactions are recorded in the normal foundry way, however we make use of alloy's `OtherFields` struct to encode additional `EIP-712` fields like `factory_deps` and `poaymaster_params` under the key specified by `ZKSYNC_TRANSACTION_OTHER_FIELDS_KEY` constant.

Reviewer: ```suggestion
During initial execution we execute the script just like a test, while we record individual transactions as they occur using the `vm.startBroadcast() / vm.stopBroadcast()` cheatcodes. These transactions are recorded in the normal foundry way, however we make use of alloy's `OtherFields` struct to encode additional `EIP-712` fields like `factory_deps` and `paymaster_params` under the key specified by `ZKSYNC_TRANSACTION_OTHER_FIELDS_KEY` constant.
```
